src/lib/server/cart.ts

```diff
@@ -163,6 +163,19 @@ export async function addToCartInDb(
 	} else if (product.variations?.length) {
 		throw error(400, 'error matching on variations choice');
 	}
+	if (
+		params.freeQuantity &&
+		(await collections.orders.countDocuments(
+			{
+				...userQuery(params.user),
+				'items.product._id': product._id,
+				status: 'pending'
+			},
+			{ limit: 1 }
+		))
+	) {
+		throw error(400, 'You already have a pending order for this free product: ' + product.name);
+	}
```

Comment:
I think what @Tirodem referred to is decreasing the available stock (maybe add it to a "pending" object) as soon as the order is created.

Then the `onOrderPaid` would move the stock from "pending" to "used", and on order cancel would move the stock from pending to available.